### PR TITLE
Merge 2.12.x to 2.13.x [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
```
% export MB=$(git merge-base origin/2.12.x origin/2.13.x)

% git log $MB..origin/2.12.x --graph --decorate --oneline  | tail -r
* 91bf0c78fc (retronym/backport/7671-java-ident, origin/pr/7738) [backport] Use Java rules for member lookup in .java sources
* 9e6ca45c2f Merge pull request #7738 from retronym/backport/7671-java-ident
|/
| * 65857146ca (origin/pr/7774, SethTisue/partest-1.1.9) partest 1.1.9 (was 1.1.7)
|/|
| |/
|\ \
* |   a2aa480be5 Merge pull request #7774 from SethTisue/partest-1.1.9
|/ /
| * | b37c0a42b3 (origin/pr/7786, diesalbla/glb_merge_for_comprehension_map) Small optimisation in glbNorm function.
|/| |
| |/ /
|\ \ \
* | |   a9337c2aa8 Merge pull request #7786 from diesalbla/glb_merge_for_comprehension_map
| |/ /
| * | | a996dc49eb misc minor readme tweaks
| * | | 5ad9e03fde remove inactive maintainers
| * | | 1bb9b7482c readme: reorder maintainers, add Stefan and Viktor
| * | | 6ff01aec2b (origin/pr/7778, SethTisue/maintainers-list) Add sjrd to the maintainer list, for interactions with Scala.js.
|/| | |
| |/ / /
|\ \ \ \
* | | |   9197882eaf Merge pull request #7778 from SethTisue/maintainers-list
| |/ / /
| * | | | e22e12da28 (retronym/topic/unblock-scala-meta, origin/pr/7799, topic/unblock-scala-meta) Restore API in typer used by scala-meta
|\ \ \ \
* | | |   ac535fdbbe Merge pull request #7799 from retronym/topic/unblock-scala-meta
| |/ / /
| * | | | ecfa631545 (origin/pr/7791, diesalbla/glbNorm_performance_tweaks) Small performance tweak to glbNorm.
|\ \ \ \
* | | |   5915f41f5f Merge pull request #7791 from diesalbla/glbNorm_performance_tweaks
| * | | 9a04c4d9b7 (origin/pr/7797, diesalbla/complexity_map_into_fold) Complexity: fold map into the sum.
|\ \ \
* | |   b63ccf2296 Merge pull request #7797 from diesalbla/complexity_map_into_fold
| |/ /
| * | | 6a8177b4e1 [backport] Upgrade to ASM 7
| * | | 1220d3c915 (origin/pr/7780, SethTisue/asm7) [backport] Nicer branch-sensitive nullness
|\ \ \
* | |   ce9a857920 (retronym/2.12.x, topic/2.12.x-backports) Merge pull request #7780 from SethTisue/asm7
| * | a1539c6934 (origin/pr/7788, diesalbla/backports_upperBound_lowerBound) Back-ports changes to add upperBound, lowerBound.
|/| |
| |_|/
|\ \ \
* | |   f83d073843 Merge pull request #7788 from diesalbla/backports_upperBound_lowerBound
| * | cfbb2a2d16 [backport] Resource management for macro/plugin classloaders, classpath JARs
| * | 8efcb4236b unpickler cycle avoidance
| * | a4d9599f1d Utility to abstract over JAR and directory I/O
| * | 69d3c00027 Utility to extract Pickles from Scala compiled classes
| * | b066d7e640 Compute pickles for .java sources
| * | 22eae50c8f .sig file support
| * | 27a976c084 Experimental support for outline typing
| * | c6e290ad4a Driver to build a set of compile @argsfile-s in pipeline.
| * | 8e58ea0bc6 Valid URL for VirtualDirectory classpath
| * | 141a72f6fa Add missing copyright headers
| * | 7707a763fc [nomerge] Restore findMacroClassloader into Analyzer for the 2.12.x series
| * | 0b1974c8e7 Refactor PipelineMain
| * | d9b98b8d74 Whitelist some binary changes to internals of scala-reflect
| * | 2c6a4b22e4 Remove outline typechecking for now
| * | 3e500b21a5 Use file size as part of the cache invalidation
| | |/
| |\ \
| * \   5747be5840 (retronym/topic/pipeline-trace-macro-omnibus-rebase, origin/pr/7712, topic/pipeline-trace-macro-omnibus-rebase) Merge remote-tracking branch 'origin/2.12.x' into topic/pipeline-trace-macro-omnibus-rebase
|\ \
* |   2961eda885 Merge pull request #7712 from retronym/topic/pipeline-trace-macro-omnibus-rebase
| |/
| * | ea0e5c8daf (origin/pr/7810, diesalbla/remove_bothnames) Remove the bothNames method from the Name class.
|\ \
* |   2a395754ce Merge pull request #7810 from diesalbla/remove_bothnames
| * 693f3a724b (origin/pr/7808, diesalbla/tryTypedApply_flatmap_exists) Typers: merge treesInResult with errorInResult
|\
*   29c6d99f28 (diesalbla/2.12.x, 2.12.x) Merge pull request #7808 from diesalbla/tryTypedApply_flatmap_exists
|/
| * 3e0ab870ce (retronym/topic/cache-policy-fix, origin/pr/7813, topic/cache-policy-fix) Honour CachePolicy.None in classloader/classpath caching
|\
*   88ed07f080 (origin/2.12.x) Merge pull request #7813 from retronym/topic/cache-policy-fix

% git merge -sours a2aa480be5
Merge made by the 'ours' strategy.

% git merge b63ccf2296
Auto-merging src/reflect/scala/reflect/internal/tpe/GlbLubs.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Typers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Implicits.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/typechecker/Implicits.scala
Auto-merging README.md
Automatic merge failed; fix conflicts and then commit the result.

% git status
On branch merge/2.12.x-to-2.12.x-20180305
All conflicts fixed but you are still merging.

Changes to be committed:
    modified:   README.md
    modified:   src/compiler/scala/tools/nsc/typechecker/Implicits.scala
    modified:   src/compiler/scala/tools/nsc/typechecker/Typers.scala
    modified:   src/reflect/scala/reflect/internal/tpe/GlbLubs.scala


% git add .

% git commit
[merge/2.12.x-to-2.12.x-20180305 98f0002726] Merge commit 'b63ccf2296' into merge/2.12.x-to-2.12.x-20180305

% git merge -sours f83d073843
Merge made by the 'ours' strategy.

% git merge -sours cfbb2a2d16
Merge made by the 'ours' strategy.

% git merge 141a72f6fa
Auto-merging src/reflect/scala/reflect/internal/pickling/UnPickler.scala
Auto-merging src/reflect/scala/reflect/internal/Definitions.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Typers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
Auto-merging src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala
Auto-merging src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala
Auto-merging src/compiler/scala/tools/nsc/symtab/classfile/AbstractFileReader.scala
Auto-merging src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
Auto-merging src/compiler/scala/tools/nsc/classpath/VirtualDirectoryClassPath.scala
Auto-merging src/compiler/scala/tools/nsc/Global.scala
Auto-merging src/compiler/scala/tools/nsc/CompilationUnits.scala
Automatic merge failed; fix conflicts and then commit the result.

% git status
gOn branch merge/2.12.x-to-2.12.x-20180305
All conflicts fixed but you are still merging.

Changes to be committed:
    modified:   src/compiler/scala/tools/nsc/CompilationUnits.scala
    modified:   src/compiler/scala/tools/nsc/Global.scala
    new file:   src/compiler/scala/tools/nsc/PickleExtractor.scala
    new file:   src/compiler/scala/tools/nsc/PipelineMain.scala
    modified:   src/compiler/scala/tools/nsc/classpath/FileUtils.scala
    modified:   src/compiler/scala/tools/nsc/classpath/VirtualDirectoryClassPath.scala
    modified:   src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
    modified:   src/compiler/scala/tools/nsc/symtab/classfile/AbstractFileReader.scala
    modified:   src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala
    modified:   src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala
    modified:   src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
    modified:   src/compiler/scala/tools/nsc/typechecker/Typers.scala
    modified:   src/reflect/scala/reflect/internal/Definitions.scala
    modified:   src/reflect/scala/reflect/internal/pickling/UnPickler.scala
    new file:   src/reflect/scala/reflect/io/RootPath.scala


% git add .

% git commit
[merge/2.12.x-to-2.12.x-20180305 aa3ffb71c3] Merge commit '141a72f6fa' into merge/2.12.x-to-2.12.x-20180305

% git merge -sours 7707a763fc
Merge made by the 'ours' strategy.

% git merge 88ed07f080
Auto-merging src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala
Auto-merging src/reflect/scala/reflect/internal/Names.scala
CONFLICT (modify/delete): src/reflect/mima-filters/2.12.0.forwards.excludes deleted in HEAD and modified in 88ed07f080. Version 88ed07f080 of src/reflect/mima-filters/2.12.0.forwards.excludes left in tree.
CONFLICT (modify/delete): src/reflect/mima-filters/2.12.0.backwards.excludes deleted in HEAD and modified in 88ed07f080. Version 88ed07f080 of src/reflect/mima-filters/2.12.0.backwards.excludes left in tree.
Auto-merging src/compiler/scala/tools/nsc/typechecker/Typers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Namers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
Auto-merging src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala
Automatic merge failed; fix conflicts and then commit the result.

% git commit
[merge/2.12.x-to-2.12.x-20180305 14ad07234a] Merge commit '88ed07f080' into merge/2.12.x-to-2.12.x-20180305


```

--------------

```
$ git log 88ed07f080..origin/2.12.x --oneline --graph --decorate | tail -r
* 57277be9a7 (origin/pr/7816, diesalbla/fix_add_deprecation_bothnames) Restore deprecated bothnames method.
* 9862ab7a15 Merge pull request #7816 from diesalbla/fix_add_deprecation_bothnames
|/
| * d62f26c982 (retronym/topic/closeable-registry, origin/pr/7825, topic/closeable-registry) Avoid NPE during global initialization under -verbose/-Ylogcp
|\
*   f0936e05ab Merge pull request #7825 from retronym/topic/closeable-registry
|/
| * ca8c69da8e (origin/pr/7822, SethTisue/pathresolver-backcompat) restore compat for PathResolver, ClassPathFactory constructors
|/|
| |/
|\ \
* |   7ae699585e Merge pull request #7822 from SethTisue/pathresolver-backcompat
|  /
| * | 1494b64d74 Avoid the parentSymbols method: less allocations.
| * | a789144411 (origin/pr/7823, diesalbla/parentsymbols_remove) Add parentSymbolsIterator method. Use it instead of the info.
|\ \
* |   25c7d20ce5 Merge pull request #7823 from diesalbla/parentsymbols_remove
|  /
| * | 9fcb2f34fc (origin/pr/7824, diesalbla/typers_avoid_flatten) Avoid using List.flatten
|\ \
* |   2b9aa1c264 (retronym/2.12.x) Merge pull request #7824 from diesalbla/typers_avoid_flatten
| * 8763166e8f (retronym/topic/picklewriter, origin/pr/7828, topic/picklewriter) Avoid trailing zero bytes in .sig files written in PipelineMain
|/|
| |/
|\ \
* |   8e71ae3348 Merge pull request #7828 from retronym/topic/picklewriter
|/ /
| * | fcb8ff7a29 (origin/pr/7837, diesalbla/copy_method_completer_replace_map2_with_foreach2) copyMethodCompleter: Replace map2 with foreach2
|/| |
| |/ /
|\ \ \
* | |   6b1fdbb7de Merge pull request #7837 from diesalbla/copy_method_completer_replace_map2_with_foreach2
| |/ /
| * | | 524c672ef2 (origin/pr/7838, diesalbla/specialize_types_avoid_zip) Avoid using List.zip in the SpecializeTypes unification
|\ \ \
* | |   56d29c4a80 Merge pull request #7838 from diesalbla/specialize_types_avoid_zip
| |/ /
| * | | 57c4888f08 Remove use of List.map in the "varianceInTypes" function.
| * | | 150e0fe350 (retronym/review/7835) Reduce SAM lambda allocation within varianceInTypes
| * | | e6cc02db5d (origin/pr/7835, diesalbla/variances_remove_listmap) Make varianceInType allocation free with a reusable instance
|\ \ \
* | |   2a18ab80a7 Merge pull request #7835 from diesalbla/variances_remove_listmap
| |/ /
| * | | 3cc07a4955 (origin/pr/7827, diesalbla/uncurry_avoid_flatten) Uncurry: avoid using List.flatten
|\ \ \
* | |   e214fe3577 Merge pull request #7827 from diesalbla/uncurry_avoid_flatten
| * | a181111338 (origin/pr/7809, diesalbla/avoid_list_creation_) Avoid the call to BaseTypeSeq.toList in type Unification.
|\ \
* |   efb9faf7f6 Merge pull request #7809 from diesalbla/avoid_list_creation_
|  /
| * | 4e718b3359 (origin/pr/7820, diesalbla/dedup_no_lists_backport) Selectively Backport some changes
|/| |
| |/ /
|\ \ \
* | |   f6930376f5 Merge pull request #7820 from diesalbla/dedup_no_lists_backport
|  / /
| * | | ca3e491114 Avoid using List flatMap withi normalizeImpl
| * | | d27a93af6e (origin/pr/7826, diesalbla/types_normalize_no_flatmap) Use LinkedHashSet instead of ListBuffer: better query times
|\ \ \
* | |   89f8daf77b Merge pull request #7826 from diesalbla/types_normalize_no_flatmap
| * | 33d9f6f303 (origin/pr/7840, diesalbla/typers_avoid_unzipping) Typers: by-pass the unzipped list of pairs.
|\ \
* |   c04a5dd786 Merge pull request #7840 from diesalbla/typers_avoid_unzipping
|/ /
| * | 6e42ccadda (origin/pr/7852, diesalbla/In_map_substInfo_use_foreach) Replace the uses of map for substInfo
|/| |
| |/ /
|\ \ \
* | |   e3c0c10660 Merge pull request #7852 from diesalbla/In_map_substInfo_use_foreach
| |/ /
| * | | 5319a9e1f5 (origin/pr/7855, diesalbla/merge_find_foreach_map) Superaccessors: merge several List method.
|\ \ \
* | |   3e98d20bf2 Merge pull request #7855 from diesalbla/merge_find_foreach_map
| * | b2bee0c095 (origin/pr/7854, diesalbla/foreachParamss) Optimisation: replace mapParamss with a foreachParamss
|\ \
* |   9f5d27d4a9 Merge pull request #7854 from diesalbla/foreachParamss
|/ /
| * | 288d49a068 (origin/pr/7858, diesalbla/Definitions_replace_map_distinct_check_singleton_with_head_tail_forall) Definitions - SymbolSet class: deforest middle list.
|/| |
| |/ /
|\ \ \
* | |   724aac58d7 Merge pull request #7858 from diesalbla/Definitions_replace_map_distinct_check_singleton_with_head_tail_forall
| |/ /
| * | | 9a41c4cd6f (origin/pr/7868, diesalbla/Checkable_avoid_allocations) Remove List Allocations in the Checkable
|\ \ \
* | |   02d7a464dc Merge pull request #7868 from diesalbla/Checkable_avoid_allocations
| * | 6a58cce5a0 (origin/pr/7865) Optimisations: avoid List allocations in the RefChecks.
|\ \
* |   6f4ee4a51d Merge pull request #7865 from diesalbla/RefChecks_avoid_flatten_maps_filter
|/ /
| * | afe6d85ea9 (origin/pr/7872) [backport] Prefer HTTPS over HTTP for pom.xml data
|\ \
* |   4acff24d35 Merge pull request #7872 from ennru/prefer-https-2.12
| * e9bd65d7f8 Fix completion of explicitly imported type names
| * 55ba885d2a (retronym/topic/completion-new, origin/pr/7833, topic/completion-new) Show constructor def strings TAB,TAB, as we do for methods
|/|
| |/
|\ \
* |   d708101d83 Merge pull request #7833 from retronym/topic/completion-new
| * 0bd07f9f8c (origin/pr/7881, diesalbla/zipped_map_convert_to_map2) Replace "zipped" by combinators without list allocations.
|/|
| |/
|\ \
* |   01cb722a28 Merge pull request #7881 from diesalbla/zipped_map_convert_to_map2
| |/
| * | c5a69fdb50 Don't give unhelpful unemptiness advice.
| * | b87e2c5b3b (origin/pr/7898, hrhino/topic/not-unempty) mailmap me!
|\ \
* |   f967dcd9f0 (diesalbla/2.12.x) Merge pull request #7898 from hrhino/topic/not-unempty
| * 98e59b41c9 (retronym/topic/implicit-shadower-reusable, origin/pr/7890) Recycle the hashset used for implicit shadowing
|/|
| |/
|\ \
* |   eb8280dbd2 (2.12.x) Merge pull request #7890 from retronym/topic/implicit-shadower-reusable
| * cad96e244b Refactor adjustTypeArgs, giving its result record a name
| * 5ba6776250 (retronym/topic/adjusted-args, origin/pr/7900, topic/adjusted-args) Avoid redundant computation of undetParams / refacator to avoid confusing similarly named idents
|\
*   6990ee369a (origin/2.12.x) Merge pull request #7900 from retronym/topic/adjusted-args
```

```
% git merge -sours 9862ab7a15
%  git merge efb9faf7f6
%  git merge -sours f6930376f5
%  git merge 6f4ee4a51d
%  git merge -sours 4acff24d35
%  git merge 6990ee369a
``

### CODE REVIEW

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 1]
```diff
@@ -257,8 +257,10 @@ trait ScalaSettings extends AbsScalaSettings
   val YdisableFlatCpCaching  = BooleanSetting    ("-YdisableFlatCpCaching", "Do not cache flat classpath representation of classpath elements from jars across compiler instances.")
   val YcachePluginClassLoader  = CachePolicy.setting("plugin", "compiler plugins")
   val YcacheMacroClassLoader   = CachePolicy.setting("macro", "macros")
-
   val YmacroClasspath = PathSetting       ("-Ymacro-classpath", "The classpath used to reflectively load macro implementations, default is the compilation classpath.", "")
+  val YpartialUnification = BooleanSetting ("-Ypartial-unification", "Enable partial unification in type constructor inference")
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: hmm, this shouldn't be revived

**Comment by retronym** [comment]
I'm looking at this PR now. Will submit an updated version tomorrow that resolves new conflicts and corrects the merge error.

**Comment by retronym** [comment]
- New conflicts with 2.13.x resolved
 - `-Ypartial-unificaction` merge errror fixes
 - New commits from 2.12.x included in this merge

**Comment by retronym** [comment]
@adriaanm This is ready now.

**Comment by adriaanm** [comment]
I still see 

```diff
diff --git c/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala w/src/compiler/scala/tools/nsc/settings/ScalaSettings.s
cala
index fa021cc..662dde6 100644
--- c/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
+++ w/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
@@ -257,8 +257,10 @@ trait ScalaSettings extends AbsScalaSettings
   val YdisableFlatCpCaching  = BooleanSetting    ("-YdisableFlatCpCaching", "Do not cache flat classpath representation of cl
asspath elements from jars across compiler instances.")
   val YcachePluginClassLoader  = CachePolicy.setting("plugin", "compiler plugins")
   val YcacheMacroClassLoader   = CachePolicy.setting("macro", "macros")
-
   val YmacroClasspath = PathSetting       ("-Ymacro-classpath", "The classpath used to reflectively load macro implementation
s, default is the compilation classpath.", "")
+  val YpartialUnification = BooleanSetting ("-Ypartial-unification", "Enable partial unification in type constructor inference")
+  val Yvirtpatmat     = BooleanSetting    ("-Yvirtpatmat", "Enable pattern matcher virtualization")
```

**Comment by adriaanm** [comment]
I'll look into the curious tenacity of those -Y flags this morning.

**Comment by retronym** [comment]
Egads.  I've hit them with my heftiest sledgehammer. `-Youtline` is gone too, it was for an experimental part of #7712 that I've deferred.

**Pull request merged**